test/stacktraces.jl

```diff
@@ -0,0 +1,79 @@
+# This file is a part of Julia. License is MIT: http://julialang.org/license
+
+# Tests for /base/stacktraces.jl
+
+# Some tests don't currently work for Appveyor 32-bit Windows
+const APPVEYOR_WIN32 = (
+    OS_NAME == :Windows && WORD_SIZE == 32 && get(ENV, "APPVEYOR", "False") == "True"
+)
+
+if !APPVEYOR_WIN32
+    let
+        @noinline child() = stacktrace()
+        @noinline parent() = child()
+        @noinline grandparent() = parent()
+        line_numbers = @__LINE__ - [3, 2, 1]
+
+        # Basic tests.
+        stack = grandparent()
+        @assert length(stack) >= 3 "Compiler has unexpectedly inlined functions"
+        @test [:child, :parent, :grandparent] == [f.func for f in stack[1:3]]
+        for (line, frame) in zip(line_numbers, stack[1:3])
+            @test [Symbol(@__FILE__), line] in
+                ([frame.file, frame.line], [frame.inlined_file, frame.inlined_line])
+        end
+        @test [false, false, false] == [f.from_c for f in stack[1:3]]
+
+        # Test remove_frames!
+        stack = StackTraces.remove_frames!(grandparent(), :parent)
+        @test stack[1] == StackFrame(:grandparent, @__FILE__, line_numbers[3])
+
+        stack = StackTraces.remove_frames!(grandparent(), [:child, :something_nonexistent])
+        @test stack[1:2] == [
+            StackFrame(:parent, @__FILE__, line_numbers[2]),
+            StackFrame(:grandparent, @__FILE__, line_numbers[3])
+        ]
+    end
+end
+
+let
+    # Test from_c
+    default, with_c, without_c = stacktrace(), stacktrace(true), stacktrace(false)
+    @test default == without_c
+    @test length(with_c) > length(without_c)
+    @test !isempty(filter(frame -> frame.from_c, with_c))
+    @test isempty(filter(frame -> frame.from_c, without_c))
+end
+
+@test StackTraces.lookup(C_NULL) == StackTraces.UNKNOWN
+
+let
+    # No errors should mean nothing in catch_backtrace
+    @test catch_backtrace() == StackFrame[]
+
+    @noinline bad_function() = throw(UndefVarError(:nonexistent))
+    @noinline function try_stacktrace()
+        try
+            bad_function()
+        catch
+            return stacktrace()
+        end
+    end
+    @noinline function try_catch()
+        try
+            bad_function()
+        catch
+            return catch_stacktrace()
+        end
+    end
+    line_numbers = @__LINE__ .- [15, 10, 5]
+
+    # Test try...catch with stacktrace
+    @test try_stacktrace()[1] == StackFrame(:try_stacktrace, @__FILE__, line_numbers[2])
```

Comment:
I believe that only one issue remains unaddressed. As mentioned [earlier](https://github.com/JuliaLang/julia/pull/14469#issuecomment-170730196), there are some inconsistencies with the way `backtrace` reports line numbers, which cause the test case on line 72 (and the corresponding doctest) to fail. Despite the `@noinline`s, some sort of optimization seems to be occurring which causes a line number in the `try` block to be reported for code run in the `catch` (or perhaps something else entirely is going on; this is just a guess).

The test case could be modified to succeed by replacing lines 54 through 69 with something like:

```julia
    @noinline good_function() = 15 / pi
    @noinline bad_function() = throw(UndefVarError(:nonexistent))
    @noinline function try_stacktrace()
        try
            good_function()
            bad_function()
        catch
            good_function()
            return stacktrace()
        end
    end
    @noinline function try_catch()
        try
            good_function()
            bad_function()
        catch
            good_function()
            return catch_stacktrace()
        end
    end
    line_numbers = @__LINE__ .- [19, 12, 6]
```

For reasons that I do not fully understand (see my [earlier comment](https://github.com/JuliaLang/julia/pull/14469#issuecomment-170730196)) this causes the line numbers to be reported correctly by `backtrace`—but I modifying the test case in this way seems like sidestepping the issue, and the tests *ought to* pass as they are.